src/dwarf_parser.cpp

```diff
@@ -1,250 +1,216 @@
+#include <atomic>
+#include <lldb/API/LLDB.h>
+#include <llvm/Config/llvm-config.h>
+#include <memory>
+#include <string>
 
 namespace bpftrace {
 
-struct FuncInfo {
-  std::string name;
-  Dwarf_Die die;
-};
+std::atomic<size_t> Dwarf::instance_count = 0;
 
-Dwarf::Dwarf(BPFtrace *bpftrace, const std::string &file_path)
-    : bpftrace_(bpftrace), file_path_(file_path)
+Dwarf::Dwarf(BPFtrace *bpftrace, std::string file_path)
+    : bpftrace_(bpftrace), file_path_(std::move(file_path))
 {
-  callbacks.find_debuginfo = dwfl_standard_find_debuginfo;
-  callbacks.section_address = dwfl_offline_section_address;
-  callbacks.debuginfo_path = NULL;
-  dwfl = dwfl_begin(&callbacks);
-  dwfl_report_offline(dwfl, file_path.c_str(), file_path.c_str(), -1);
-  dwfl_report_end(dwfl, NULL, NULL);
-}
+  if (instance_count++ == 0)
+    lldb::SBDebugger::Initialize();
 
-std::unique_ptr<Dwarf> Dwarf::GetFromBinary(BPFtrace *bpftrace,
-                                            const std::string &file_path)
-{
-  std::unique_ptr<Dwarf> dwarf(new Dwarf(bpftrace, file_path));
-  Dwarf_Addr bias;
-  if (dwfl_nextcu(dwarf->dwfl, NULL, &bias) == NULL)
-    return nullptr;
+  debugger_ = lldb::SBDebugger::Create(false);
 
-  return dwarf;
+  lldb::SBError error;
+  target_ = debugger_.CreateTarget(
+      file_path_.c_str(), nullptr, nullptr, false, error);
+  if (!error.Success() || !target_.IsValid()) {
+    throw error;
+  }
 }
 
 Dwarf::~Dwarf()
 {
-  dwfl_end(dwfl);
+  if (--instance_count == 0)
+    lldb::SBDebugger::Terminate();
 }
 
-static int get_func_die_cb(Dwarf_Die *func_die, void *arg)
+std::unique_ptr<Dwarf> Dwarf::GetFromBinary(BPFtrace *bpftrace,
+                                            std::string file_path)
 {
-  auto *func_info = static_cast<struct FuncInfo *>(arg);
-  if (dwarf_hasattr(func_die, DW_AT_name) &&
-      dwarf_diename(func_die) == func_info->name) {
-    func_info->die = *func_die;
-    return DWARF_CB_ABORT;
+  try {
+    return std::unique_ptr<Dwarf>(new Dwarf(bpftrace, std::move(file_path)));
+  } catch (const lldb::SBError &error) {
+    LOG(ERROR) << "Failed to parse DebugInfo: " << error.GetCString();
+    return nullptr;
   }
-  return DWARF_CB_OK;
 }
 
-std::optional<Dwarf_Die> Dwarf::get_func_die(const std::string &function) const
+lldb::SBFunction Dwarf::get_function(const std::string &function)
 {
-  struct FuncInfo func_info = { .name = function, .die = {} };
+  auto functions = target_.FindFunctions(function.c_str());
+  if (!functions.IsValid()) {
+    return lldb::SBFunction();
+  }
 
-  Dwarf_Die *cudie = nullptr;
-  Dwarf_Addr cubias;
-  while ((cudie = dwfl_nextcu(dwfl, cudie, &cubias)) != nullptr) {
-    if (dwarf_getfuncs(cudie, get_func_die_cb, &func_info, 0) > 0)
-      return func_info.die;
+  for (uint32_t i = 0; i < functions.GetSize(); i++) {
+    auto func = functions.GetContextAtIndex(i).GetFunction();
+    if (func && func.GetName() != nullptr && func.GetName() == function) {
```

Comment:
I think that you could omit the middle part, can't you?
```suggestion
    if (func && func.GetName() == function) {
```

Reply:
We can't, `std::string` doesn't check the pointer when comparing against a `const char*`, so the check must be done beforehand to prevent a crash. However, I'm going to simplify it a bit.